File: 卡码/代码随想录/Python/动态规划/03-打家劫舍/LeetCode_0337.py
#337. 打家劫舍 III
from typing import *

class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def traversal(self,root:Optional[TreeNode])->List[int]:
    if not root:
      return [0,0]
    left:List[int]=self.traversal(root.left)
    right:List[int]=self.traversal(root.right)
    #不偷
    v0:int=max(left[0],left[1])+max(right[0],right[1])
    #偷
    v1:int=left[0]+right[0]+root.val
    return [v0,v1]
  def rob(self, root: Optional[TreeNode]) -> int:
    res:List[int]=self.traversal(root)
    return max(res[0],res[1])

File: 卡码/代码随想录/Go/动态规划/03-打家劫舍/LeetCode_0337.go
// 337. 打家劫舍 III
package main

import (
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rob(root *TreeNode) int {
	var traversal func(root *TreeNode) []int
	traversal = func(root *TreeNode) []int {
		if root == nil {
			return []int{0, 0}
		}
		left := traversal(root.Left)
		right := traversal(root.Right)
		//不偷
		v0 := int(math.Max(float64(left[0]), float64(left[1]))) + int(math.Max(float64(right[0]), float64(right[1])))
		//偷
		v1 := left[0] + right[0] + root.Val
		return []int{v0, v1}
	}
	res := traversal(root)
	return int(math.Max(float64(res[0]), float64(res[1])))
}
